fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,647 @@
+    else:
+        epoch_participation = state.previous_epoch_participation
+
+    return set([
+        i for i in active_validator_indices 
+        if has_flag(epoch_participation[i], TIMELY_TARGET_FLAG_INDEX)
+    ])
+```
+
+#### `get_ffg_support`
+
+```python
+def get_ffg_support(store: Store, block_root: Root) -> Gwei:
+    """
+    Returns the total weight supporting the checkpoint in the block's chain at block's epoch.
+    """
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+    current_epoch = get_current_store_epoch(store)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    checkpoint_root = get_checkpoint_block(store, block_root, block_epoch)
+    checkpoint = Checkpoint(root=checkpoint_root, epoch=block_epoch)
+
+    if checkpoint not in store.checkpoint_states:
+        return Gwei(0)
+
+    checkpoint_state = store.checkpoint_states[checkpoint]
+
+    leaf_roots = get_leaf_block_roots(store, block_root)
+
+    active_validator_indices = get_active_validator_indices(checkpoint_state, block_epoch)
+    participating_indices = set().union(*[
+        get_epoch_participating_indices(
+            store.block_states[root], 
+            active_validator_indices, 
+            block_epoch == current_epoch
+        )
+        for root in leaf_roots
+    ])
+
+    return get_total_balance(checkpoint_state, participating_indices)
+```
+
+
+
+#### `is_ffg_confirmed_current_epoch`
+
+```python
+def is_ffg_confirmed_current_epoch(total_active_balance: int,
+                                   ffg_support_for_checkpoint: int,
+                                   remaining_ffg_weight: int,
+                                   confirmation_byzantine_threshold: int,
+                                   confirmation_slashing_threshold: int) -> bool:
+    """
+    Helper function of `is_ffg_confirmed` dealing with the case that the block to be confirmed is
+    from the current epoch.
+    """
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            (total_active_balance * confirmation_byzantine_threshold - 1) // 100 + 1,
+            confirmation_slashing_threshold,
+            ffg_support_for_checkpoint
+        )
+    )
+
+    """
+    Returns whether the following condition is true using only integer arithmetic
+    2 / 3 * total_active_balance <= (
+        ffg_support_for_checkpoint - max_adversarial_ffg_support_for_checkpoint +
+        (1 - confirmation_byzantine_threshold / 100) * remaining_ffg_weight
+    )
+    """
+
+    return (
+        200 * total_active_balance <=
+        ffg_support_for_checkpoint * 300 + (300 - 3 * confirmation_byzantine_threshold) *
+        remaining_ffg_weight - max_adversarial_ffg_support_for_checkpoint * 300
+    )
+```
+
+#### `is_ffg_confirmed_previous_epoch`
+
+```python
+def is_ffg_confirmed_previous_epoch(total_active_balance: int,
```

Comment:
The motivation why I created two methods was because I found the reasoning about the conditions required when the block is from the previous epoch vs when the block is from the current epoch quite different, even though they lead to similar formulas.

However, I do not hold a strong opinion on whether it would be better to take a more general approach in writing the code.
Happy to go with the general consensus around this.

Reply:
> This function seems to be a special case of `is_ffg_confirmed_for_epoch` with `remaining_ffg_weight == 0`. 
You are actually right. `is_ffg_confirmed_previous_epoch` is exactly the same of `is_ffg_confirmed_current_epoch` with `remaining_ffg_weight == 0`.

The same holds for  `get_ffg_confirmation_score`

I have just pushed a commit to generalize both is_ffg_confirmed and get_ffg_confirmation_score